Examples/Algorithms/Geant4HepMC/src/EventAction.cpp

```diff
@@ -12,14 +12,119 @@
+    // between
+    HepMC3::GenVertexPtr currentVertex = particleOut->production_vertex();
+    HepMC3::GenParticlePtr nextParticle = currentVertex->particles_in()[0];
+    // Cut connections from particle and remove it
+    if (particleIn->end_vertex()) {
+      particleIn->end_vertex()->remove_particle_in(particleIn);
+    }
+    currentVertex->remove_particle_out(particleIn);
+    event.remove_particle(particleIn);
+    // Cut connections from vertext and remove it
+    particleIn = nextParticle;
+    currentVertex->remove_particle_in(particleIn);
+    event.remove_vertex(currentVertex);
+  }
+  // Walk forwards to find all vertices in a row that match the pattern
+  while (findAttribute(particleOut->end_vertex(), processFilter)) {
+    // Take the particles after the vertex and remove particles & vertices in
+    // between
+    HepMC3::GenVertexPtr currentVertex = particleOut->end_vertex();
+    HepMC3::GenParticlePtr nextParticle = currentVertex->particles_out()[0];
+    // Cut connections from particle and remove it
+    if (particleOut->production_vertex())
+      particleOut->production_vertex()->remove_particle_out(particleOut);
+    currentVertex->remove_particle_in(particleOut);
+    event.remove_particle(particleOut);
+    // Cut connections from vertext and remove it
+    particleOut = nextParticle;
+    currentVertex->remove_particle_out(particleOut);
+    event.remove_vertex(currentVertex);
+  }
+
+  // Build the dummy vertex
+  auto reducedVertex = std::make_shared<HepMC3::GenVertex>();
+  event.add_vertex(reducedVertex);
+
+  reducedVertex->add_particle_in(particleIn);
+  reducedVertex->add_particle_out(particleOut);
+
+  // Remove and replace the old vertex
+  event.remove_vertex(vertex);
+  vertex = reducedVertex;
+}
+
+/// @brief This method walks over all vertices and tests whether it should be
+/// filtered
+///
+/// @param [in, out] event The underlying event
+/// @param [in, out] The current vertex under investigation
+/// @param [in] processFilter List of processes that will be filtered
+void followOutgoingParticles(HepMC3::GenEvent& event,
+                             HepMC3::GenVertexPtr vertex,
+                             const std::vector<std::string>& processFilter) {
+  // Replace and reduce vertex if it should be filtered
+  if (findAttribute(vertex, processFilter)) {
+    reduceVertex(event, vertex, processFilter);
+  }
+  // Move forward to the next vertices
+  for (const auto& particle : vertex->particles_out()) {
+    followOutgoingParticles(event, particle->end_vertex(), processFilter);
+  }
+}
+}  // namespace
+
 ActsExamples::EventAction* ActsExamples::EventAction::s_instance = nullptr;
 
 ActsExamples::EventAction* ActsExamples::EventAction::instance() {
   // Static acces function via G4RunManager
   return s_instance;
 }
 
-ActsExamples::EventAction::EventAction() : G4UserEventAction() {
+ActsExamples::EventAction::EventAction(std::vector<std::string>& processFilter)
```

Comment:
```suggestion
ActsExamples::EventAction::EventAction(std::vector<std::string> processFilter)
```
This can not be just a reference, it has to be either by-value or by const reference. The current implementation will probably do the following: it receives the input parameter by reference, i.e. `processFilter` points to the parameter in the configuration struct. The subsequent `std::move` call moves the underlying storage from the pointed-to object (the one in the configuration struct) and reassigns to `m_processFilter`. That leaves the pointed-to object in an invalid/empty state. If we use the variable in the configuration struct more than once, two calls will see different variables.

If the parameter is received by value, i.e. `std::vector<std::string> processFilter`, calling `EventAction(cfg.processesFilter)` will automatically create a copy and that copy is then moved into the member variable. If the parameter is received by const reference, i.e. `const std::vector<std::string>& processFilter` the `EventAction(cfg.processesFilter)` will not create a copy, the parameter references the original value; no `std::move(...)` is required and setting the member variable will automatically create a copy. Both options are valid. My personal preference is the former.

Reply:
done